### PR TITLE
Correct test expectations

### PR DESCRIPTION
**Pull request by dougbu**
### Correct test expectations
- use valid `multipart/form-data` content type; include a `boundary`
- correct expectations of `FormCollection` model binder's operation
 - restore tests actually skipped for either of the above reasons
- add more tests with `ModelBindingResult.Model==null` and both `IsModelSet` values

#### Remove test references to Won't Fix bug #2473
- restore #2473 tests; update test expectations
- expect `null` composite results whenever binding fails

#### Restore test skipped due to "#2646"
- that issue does not exist; was likely #2466 or similar fixed bug

#### Minor src changes
- remove unused variable and unnecessary nesting in `DefaultControllerActionArgumentBinder`
- remove dangling mention of `[DefaultValue]` in `ComplexModelDtoModelBinder`

#### nits:
- remove empty delegates from some `GetOperationBindingContext` calls; `null` fine
- do some `using` cleanup
- combine two test methods in `KeyValuePairModelBinderTest`
- name a few more arguments

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs [hunk 1]
```diff
@@ -84,12 +84,12 @@ public void BindingSourceModelBinder_ThrowsOnNonGreedySource()
...
 
-            // Act 
+            // Act
             var result = await binder.BindModelAsync(context);
 
             // Assert
             Assert.NotNull(result);
-            Assert.True(result.IsModelSet);
+            Assert.False(result.IsModelSet);
             Assert.True(binder.WasBindModelCoreCalled);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs]: While you're doing cleanup here can you rename these tests to be more accurate? `BindingSourceModelBinder_ReturnsTrue_MatchingSource` `BindingSourceModelBinder_ReturnsFalse_NonMatchingSource`

They don't really return true or false

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs]: :ok:

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs [hunk 2]
```diff
@@ -445,6 +445,7 @@ public class CompositeModelBinderTest
                 .Returns(
                     delegate (ModelBindingContext context)
                     {
+                        // Note this ModelBindingResult is not valid.
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs]: It's not really invalid, the model value will just be ignored.

File: test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs [hunk 3]
```diff
@@ -361,7 +361,7 @@ public class ControllerActionArgumentBinderTests
             binder
                 .Setup(b => b.BindModelAsync(It.IsAny<ModelBindingContext>()))
                 .Returns(Task.FromResult(
-                    result: new ModelBindingResult(model: null, key: string.Empty, isModelSet: true)));
+                    result: new ModelBindingResult(model: null, key: string.Empty, isModelSet: false)));
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs]: This really changes the meaning of this test.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs [hunk 4]
```diff
@@ -12,89 +12,11 @@ namespace Microsoft.AspNet.Mvc.IntegrationTests
 {
     public class BinderTypeBasedModelBinderIntegrationTest
     {
-        // The NullModelBinder and NullModelNotSetModelBinder return a non null ModelBindingResult but a null model.
-        [Theory(Skip = "ModelBindingResult should be non null if a model binder returns a non null resul #2473.")]
-        [InlineData(typeof(NullModelBinder), true)]
-        [InlineData(typeof(NullModelNotSetModelBinder), false)]
-        public async Task BindParameter_WithModelBinderType_NoData(Type modelBinderType, bool isModelSet)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs]: Are these two tests dupes now?

File: test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs (hunk 3 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs]: Not really much at all. Note I didn't change the assertions at all. My thought was to avoid testing a corner case in this general test. Few if any model binders return a result with `Model == null && IsModelSet`. But I can revert this one-liner change or rename the test if you prefer.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs [hunk 5]
```diff
@@ -67,6 +69,21 @@ public static IObjectModelValidator GetObjectValidator(MvcOptions options = null
...
+            {
+                if (typeof(TModelBinder) == options.ModelBinders[index].GetType())
+                {
+                    options.ModelBinders.Insert(index + 1, new AssertIfCalledModelBinder());
+                    return;
+                }
+            }
+
+            Assert.False(true, $"Unable to find { typeof(TModelBinder).FullName } in { nameof(MvcOptions) }.");
+        }
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs]: I'm really not a fan of this in the integration tests. Why is this needed? If these tests are failing then let them fail.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs]: Mostly irrelevant since the `CompositeModelBinder` should return `null` result in this case. I could change the assertions if we care about the scenario. Looked unimportant to me but I'm fine either way.

File: test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs (hunk 3 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs]: See `DefaultControllerActionArgumentBinder`@130 - there's code in place to specifically handle this scenario, and this is a test written specifically to target that code.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs (hunk 5 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs]: This meaning the assertion? It's a guard to ensure the setup does what the test expects.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs (hunk 4 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs]: I think it would be preferrable to change the expectations to match the behavior then. It sounds like removing these tests would mean that we just don't have coverage for the scenario.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs]: I mean the whole thing `UpdateOptionsToEnsureNothingFollows` does. Why is it needed? 

If the goal is to verify the behavior of an individual model binder (returning null/not-null), we should have unit tests for that.

File: test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs (hunk 3 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs]: Then the test should confirm correct behaviour when `isModelSet` is _either_ `true` or `false`.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs]: :ok:

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs (hunk 5 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs]: The goal is to verify the system pays attention to what the individual model binders return.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs [hunk 6]
```diff
@@ -291,7 +291,7 @@ public class CompositeModelBinderTest
             var modelBinder = new Mock<IModelBinder>();
             modelBinder
                 .Setup(mb => mb.BindModelAsync(It.IsAny<ModelBindingContext>()))
-                .Returns(Task.FromResult(new ModelBindingResult(null, "someName", true)));
+                .Returns(Task.FromResult(new ModelBindingResult(model: "some value", key: "someName", isModelSet: true)));
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs]: The test says, sets null value and sets the model state key.. why are we chaing it to a non null value?

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs (hunk 2 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs]: With my latest changes, this test does not make sense. 
IMO lets replace it with `BindModel_DoesNotAddAValidationNode_IfModelBindingResultDoesNotContainAValidationNode`
PS: shorten the name if you want to.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/FormValueProviderFactoryTests.cs [hunk 7]
```diff
@@ -33,8 +33,8 @@ public void GetValueProvider_ReturnsNull_WhenContentTypeIsNotFormUrlEncoded()
         [Theory]
         [InlineData("application/x-www-form-urlencoded")]
         [InlineData("application/x-www-form-urlencoded;charset=utf-8")]
-        [InlineData("multipart/form-data")]
-        [InlineData("multipart/form-data;charset=utf-8")]
+        [InlineData("multipart/form-data; boundary=----WebKitFormBoundarymx2fSWqWSd0OxQqq")]
+        [InlineData("multipart/form-data; boundary=----WebKitFormBoundarymx2fSWqWSd0OxQqq; charset=utf-8")]
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/FormValueProviderFactoryTests.cs]: :+1:

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/KeyValuePairModelBinderTest.cs [hunk 8]
```diff
@@ -139,14 +139,13 @@ public class KeyValuePairModelBinderTest
             var innerBinder = new Mock<IModelBinder>();
             innerBinder
                 .Setup(o => o.BindModelAsync(It.IsAny<ModelBindingContext>()))
-                .Returns((ModelBindingContext mbc) =>
+                .Returns((ModelBindingContext context) =>
                 {
-                    Assert.Equal("someName.key", mbc.ModelName);
-                    return Task.FromResult(new ModelBindingResult(null, string.Empty, true));
+                    Assert.Equal("someName.key", context.ModelName);
+                    return Task.FromResult(new ModelBindingResult(model: "not int", key: string.Empty, isModelSet: true));
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/KeyValuePairModelBinderTest.cs]: line break here.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs [hunk 9]
```diff
@@ -105,7 +105,7 @@ public TestableBindingSourceModelBinder(BindingSource source)
             protected override Task<ModelBindingResult> BindModelCoreAsync([NotNull] ModelBindingContext bindingContext)
             {
                 WasBindModelCoreCalled = true;
-                return Task.FromResult(new ModelBindingResult(null, bindingContext.ModelName, true));
+                return Task.FromResult(new ModelBindingResult(model: null, key: bindingContext.ModelName, isModelSet: false));
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs]: All the changes in this class seem to be cosmetic .. have we made sure that none of the other tests depend on a behavior like model is null and IsModelSet is true?

File: test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs (hunk 4 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs]: :+1: with @rynowak I would rather have tests that ensure the behavior.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/FormFileModelBindingIntegrationTest.cs [hunk 10]
```diff
@@ -154,9 +154,8 @@ private class Address
 
             // Assert
 
-            // ModelBindingResult
-            Assert.NotNull(modelBindingResult); // Fails due to bug #2456
-            Assert.Null(modelBindingResult.Model);
+            // ModelBindingResult; expect null because binding failed.
+            Assert.Null(modelBindingResult);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/FormFileModelBindingIntegrationTest.cs]: This is wrong, the FormFileModelBinder should return non null MBR, because the type is IFormFile. That is what bug #2456 was about. Because of the quirkiness of the CompositeModelBinder, the result is null. Lets discuss this.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/FormFileModelBindingIntegrationTest.cs]: Will make sure the test name matches the expected behaviour.

Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/FormFileModelBindingIntegrationTest.cs]: Thanks, discussed in person what is here makes sense.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs (hunk 9 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs]: Not exactly cosmetic. I should have tested `isModelSet: false` with `model: null` as well as `isModelSet: true`. Instead I just switched `isModelSet` here. Will make sure both are tested.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs]: Test makes sense since no `ModelValidationNode` with `isModelSet: false` can definitely happen. Or am I missing something?

**Comment by dougbu** [comment]
Rebased and updated
- also changed PR description to match latest overall changes
- note additional fixes e.g. restoration of test skipped due to non-existent "#2646" issue and src nits

File: test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs [hunk 11]
```diff
@@ -87,14 +110,14 @@ private class Person2
...
+            Assert.Null(modelBindingResult);
 
             // ModelState
             Assert.True(modelState.IsValid);
             Assert.Empty(modelState.Keys);
         }
-        
-        // ModelBinderAttribute can be used without specifing the binder type. 
+
+        // ModelBinderAttribute can be used without specifying the binder type.
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs]: :dancers:

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs [hunk 12]
```diff
@@ -81,24 +81,21 @@ public class DefaultControllerActionArgumentBinder : IControllerActionArgumentBi
...
                     _modelMetadataProvider,
                     metadata,
                     modelBindingResult.Model);
+                var validationContext = new ModelValidationContext(
+                    modelBindingContext.BindingSource,
+                    operationContext.ValidatorProvider,
+                    modelState,
+                    modelExplorer);
 
-                if (modelBindingResult.ValidationNode != null)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs]: why is this removed? ValidationNode can be null if a model binder chooses not to use a MVN simply because it does not want to validate the model. I expected some tests to break, are they breaking?

**Comment by harshgMSFT** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs]: It was moved up into the surrounding `if`

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/ComplexModelDtoModelBinder.cs [hunk 13]
```diff
@@ -32,7 +32,7 @@ public sealed class ComplexModelDtoModelBinder : IModelBinder
                     await bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(propertyBindingContext);
                 if (modelBindingResult == null)
                 {
-                    // Could not bind. Add a result so MutableObjectModelBinder will check for [DefaultValue].
+                    // Could not bind. Let MutableObjectModelBinder know explicitly.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/ComplexModelDtoModelBinder.cs]: Not suggesting to add more to this PR, but the comment suggests this block is unnecessary. IMO correcting the comment is the right thing to do for now though :100:

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs (hunk 12 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs]: my bad. Please ignore my comment.

**Comment by harshgMSFT** [comment]
:shipit:

**Comment by rynowak** [comment]
:shipit:

**Comment by dougbu** [comment]
26795bd
